test/e2e/pv_reclaimpolicy.go

```diff
@@ -0,0 +1,158 @@
+			framework.SkipUnlessProviderIs("vsphere")
+			vsp, err := vsphere.GetVSphere()
+			Expect(err).NotTo(HaveOccurred())
+
+			volumePath, pv, pvc, err = testSetupVSpherePersistentVolumeReclaim(vsp, c, ns, v1.PersistentVolumeReclaimDelete)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("verify pv is deleted")
+			err = framework.WaitForPersistentVolumeDeleted(c, pv.Name, 3*time.Second, 300*time.Second)
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		/*
+			This test Verify persistent volume should be retained when reclaimPolicy on the PV is set to retain
+			and associated claim is deleted
+
+			Test Steps:
+			1. Create vmdk
+			2. Create PV Spec with volume path set to VMDK file created in Step-1, and PersistentVolumeReclaimPolicy is set to Retain
+			3. Create PVC with the storage request set to PV's storage capacity.
+			4. Wait for PV and PVC to bound.
+			5. Delete PVC
+			6. Verify PV is retained.
+			7. Delete PV.
+		*/
+
+		It("should retain persistent volume when reclaimPolicy set to retain when associated claim is deleted", func() {
+			framework.SkipUnlessProviderIs("vsphere")
+			vsp, err := vsphere.GetVSphere()
+			Expect(err).NotTo(HaveOccurred())
+
+			volumePath, pv, pvc, err = testSetupVSpherePersistentVolumeReclaim(vsp, c, ns, v1.PersistentVolumeReclaimRetain)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("verify pv is retained")
+			framework.Logf("Waiting for PV %v to become Released", pv.Name)
+			err = framework.WaitForPersistentVolumePhase(v1.VolumeReleased, c, pv.Name, 3*time.Second, 300*time.Second)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("delete pv")
+			deletePersistentVolume(c, pv.Name)
+			err = framework.WaitForPersistentVolumeDeleted(c, pv.Name, 3*time.Second, 300*time.Second)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("delete volume")
+			err = vsp.DeleteVolume(volumePath)
+			Expect(err).NotTo(HaveOccurred())
+		})
+	})
+})
+
+// Test Setup for persistentvolumereclaim tests for vSphere Provider
+func testSetupVSpherePersistentVolumeReclaim(vsp *vsphere.VSphere, c clientset.Interface, ns string, persistentVolumeReclaimPolicy v1.PersistentVolumeReclaimPolicy) (volumePath string, pv *v1.PersistentVolume, pvc *v1.PersistentVolumeClaim, err error) {
+	By("running testSetupVSpherePersistentVolumeReclaim")
+	By("creating vmdk")
+	volumePath, err = createVSphereVolume(vsp, nil)
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	By("creating the pv")
+	pv = getVSpherePersistentVolumeSpec(volumePath, persistentVolumeReclaimPolicy, nil)
+	pv, err = c.CoreV1().PersistentVolumes().Create(pv)
+	if err != nil {
+		return volumePath, nil, nil, err
+	}
+
+	By("creating the pvc")
+	pvc = getVSpherePersistentVolumeClaimSpec(ns, nil)
+	pvc, err = c.CoreV1().PersistentVolumeClaims(ns).Create(pvc)
+	if err != nil {
+		return volumePath, pv, nil, err
+	}
+
+	By("wait for the pv and pvc")
+	waitOnPVandPVC(c, ns, pv, pvc)
+
+	By("delete pvc")
+	deletePersistentVolumeClaim(c, pvc.Name, ns)
+	return volumePath, pv, pvc, nil
```

Comment:
I am not sure what the kubernetes convention is for returning named values? Go permits you to just do a simple `return` and it will return the assigned return values or their defaults.